src/mcts/storage.cc

```diff
@@ -3,44 +3,51 @@
 
 namespace mcts {
 
-static std::mutex freeStoragesMutex;
-static std::list<Storage*> freeStorages;
+static std::mutex storagesMutex;
+static std::vector<std::unique_ptr<Storage>> allStorages;
+static std::vector<Storage*> freeStorages;
+
 
 Node* Storage::newNode() {
-  if (chunkIndex >= chunks.size()) {
-    Node* newChunk = (Node*)std::aligned_alloc(128, sizeof(Node) * chunkSize);
-    new (newChunk) Node[chunkSize];
-    for (size_t i = 0; i != chunkSize; ++i) {
-      newChunk[i].setStorageAndId(this, i);
+  if (chunks_it == chunks.end()) {
+    chunks.emplace_back();
+    chunks_it = std::prev(chunks.end());
+
+    chunks_it->reserve(CHUNK_SIZE);
+    for (size_t i = 0; i < CHUNK_SIZE; ++i) {
+      chunks_it->emplace_back(this);
     }
-    chunks.push_back(newChunk);
   }
-  Node* r = chunks[chunkIndex] + subIndex;
+
+  Node* r = &(*chunks_it)[subIndex];
   ++subIndex;
-  if (subIndex == chunkSize) {
+  if (subIndex == chunks_it->size()) {
     subIndex = 0;
-    ++chunkIndex;
+    ++chunks_it;
   }
+
   ++allocated;
   return r;
 }
 
 void Storage::freeNode(Node* node) {
+  assert(node->storage_ == this);
+
   --allocated;
   assert(allocated >= 0);
   if (allocated == 0) {
-    chunkIndex = 0;
+    chunks_it = chunks.begin();
     subIndex = 0;
-    std::lock_guard l(freeStoragesMutex);
+    std::lock_guard l(storagesMutex);
     freeStorages.push_back(this);
   }
 }
 
 Storage* Storage::getStorage() {
-  std::unique_lock l(freeStoragesMutex);
+  std::lock_guard l(storagesMutex);
```

Comment:
`lock_guard` cannot be freed by `unlock` -- we don't do that anymore.

Reply:
Let's talk about this!

Reply:
They seem to be basically the same except `unique_lock` can be released by calling `.release()` and then acquired again, whereas `lock_guard` isn't released unless it goes out of scope. Which is more predictable, so if possible it's better to use it.

See [lock_guard vs unique_lock](https://stackoverflow.com/a/20516876) and [why lock_guard isntead of scoped_lock for a single mutex](https://stackoverflow.com/a/60172828).